toqito/state_props/is_unextendible_product_basis.py

```diff
@@ -0,0 +1,117 @@
+    :math:`S` of pure orthogonal product states spanning a proper subspace :math:`\mathcal{H}_S` of :math:`\mathcal{H}`.
+    An unextendible product basis (UPB) is a PB whose complementary subspace :math:`\mathcal{H}_S-\mathcal{H}` contains
+    no product state.  This function is inspired from `IsUPB <https://qetlab.com/IsUPB>`_ in QETLAB.
+
+    Examples
+    ==========
+    See :func:`toqito.states.tile`. All the states together form a UPB:
+    >>> import numpy as np
+    >>> from toqito.states import tile
+    >>> from toqito.state_props import is_unextendible_product_basis
+    >>>
+    >>> upb_tiles = np.array([tile(i) for i in range(5)])
+    >>> dims = np.array([3, 3])
+    >>> is_unextendible_product_basis(upb_tiles, dims)
+    (True, None)
+
+    However, the first 4 do not:
+
+    >>> import numpy as np
+    >>> from toqito.states import tile
+    >>> from toqito.state_props import is_unextendible_product_basis
+    >>>
+    >>> non_upb_tiles = np.array([tile(i) for i in range(4)])
+    >>> dims = np.array([3, 3])
+    >>> is_unextendible_product_basis(non_upb_tiles, dims)
+    (False, array([-0.00000000e+00,  0.00000000e+00,  0.00000000e+00, -0.00000000e+00,
+        0.00000000e+00,  0.00000000e+00, -1.11022302e-16,  7.07106781e-01,
+        7.07106781e-01]))
+
+    The orthogonal state is given by
+
+    .. math::
+        \frac{1}{\sqrt{2}} |2\rangle \left( |1\rangle + |2\rangle \right)
+
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :raises ValueError: If product of dimensions does not match the size of a vector.
+    :raises ValueError: If at least one vector is not a product state.
+    :param vecs: The list of states.
+    :param dims: The list of dimensions.
+    :return: Returns a tuple. The first element is :code:`True` if input is a UPB and :code:`False` otherwise. The
+             second element is a witness (a product state orthogonal to all the input vectors) if the input is a
+             PB and :code:`None` otherwise.
+    """
+    # Some error handling:
+    if isinstance(vecs, list):
+        vecs = np.array(vecs)
+    if isinstance(dims, list):
+        dims = np.array(dims)
+
+    if np.prod(dims) != vecs.shape[1]:
+        raise ValueError("Product of dimensions does not equal the size of each vector")
+
+    if not all(is_product(vec, dims)[0] for vec in vecs):
+        raise ValueError("At least one vector is not a product state")
+
+    # Number of parties (m).
+    num_parties = dims.shape[0]
+
+    # Number of vectors (n).
+    num_vecs = vecs.shape[0]
+
+    # If n<m vectors are provided, then we cannot generate set partitions,
+    # so it is not a UPB. We will extend the set with m-n null vectors and
+    # run the same algorithm.
+    if num_vecs < num_parties:
+        vecs = np.append(vecs, np.zeros(shape=(num_parties - num_vecs, *vecs.shape[1:])), axis=0)
+        num_vecs = vecs.shape[0]
```

Comment:
Pylint flagged this. 


toqito/state_props/is_unextendible_product_basis.py:85:7: R6103: Use 'if (num_vecs := vecs.shape[0]) < num_parties:' instead (consider-using-assignment-expr)